File: src/app/useCases/trafficSource/deleteTrafficSourceUseCase.ts
import { HttpAdapter } from "../../../infra/adapters/httpAdapter";
import { TrafficSourceRepository } from "../../../domain/repositories/trafficSource";

class DeleteTrafficSourceUseCase {
  constructor(private trafficSourceRepository: TrafficSourceRepository) {}

  async execute(trafficSourceId: string, userId: string) {
    const trafficSource =
      await this.trafficSourceRepository.findById(trafficSourceId);

    if (!trafficSource) {
      throw HttpAdapter.notFound("Traffic source not found");
    }

    await this.trafficSourceRepository.deleteTrafficSource(trafficSource.id);
  }
}

export { DeleteTrafficSourceUseCase };

File: internal/app/useCases/trafficSource/deleteTrafficSourceUseCase.go
package trafficSourceUseCases

import (
	domainErrors "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/errors"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/repositories"
)

type DeleteTrafficSourceUseCase struct {
	trafficSourceRepository repositories.TrafficSourceRepository
}

func NewDeleteTrafficSourceUseCase(
	trafficSourceRepository repositories.TrafficSourceRepository,
) *DeleteTrafficSourceUseCase {
	return &DeleteTrafficSourceUseCase{trafficSourceRepository: trafficSourceRepository}
}

func (t *DeleteTrafficSourceUseCase) Handle(userId string, trafficSourceId string) error {
	trafficSource, err := t.trafficSourceRepository.FindById(trafficSourceId)

	if err != nil {
		return err
	}

	if trafficSource.UserId != userId {
		return domainErrors.Invalid("You aren't traffic source owner.")
	}

	err = t.trafficSourceRepository.Delete(trafficSource)

	if err != nil {
		return err
	}

	return nil
}
